# Tuning SGDRegressor

In [1]:
import numpy
import csv
import time
import pandas
import signal
import warnings
warnings.filterwarnings("ignore")

In [2]:
from sklearn.metrics import make_scorer
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_validate
from sklearn.pipeline import Pipeline

In [3]:
from sklearn.multioutput import ClassifierChain
from sklearn.multioutput import MultiOutputRegressor
from sklearn.multioutput import MultiOutputClassifier

from sklearn.ensemble.weight_boosting import AdaBoostClassifier
from sklearn.ensemble.weight_boosting import AdaBoostRegressor
from sklearn.ensemble.bagging import BaggingClassifier
from sklearn.ensemble.bagging import BaggingRegressor
from sklearn.ensemble.forest import ExtraTreesClassifier
from sklearn.ensemble.forest import ExtraTreesRegressor
from sklearn.ensemble.gradient_boosting import GradientBoostingClassifier
from sklearn.ensemble.gradient_boosting import GradientBoostingRegressor
from sklearn.ensemble.forest import RandomForestClassifier
from sklearn.ensemble.forest import RandomForestRegressor

from sklearn.linear_model.bayes import ARDRegression
from sklearn.linear_model.bayes import BayesianRidge
from sklearn.naive_bayes import BernoulliNB
from sklearn.cross_decomposition.cca_ import CCA
from sklearn.tree.tree import DecisionTreeClassifier
from sklearn.tree.tree import DecisionTreeRegressor
from sklearn.linear_model.coordinate_descent import ElasticNet
from sklearn.tree.tree import ExtraTreeClassifier
from sklearn.tree.tree import ExtraTreeRegressor
from sklearn.naive_bayes import GaussianNB
from sklearn.gaussian_process.gpc import GaussianProcessClassifier
from sklearn.gaussian_process.gpr import GaussianProcessRegressor
from sklearn.linear_model.huber import HuberRegressor
from sklearn.neighbors.classification import KNeighborsClassifier
from sklearn.neighbors.regression import KNeighborsRegressor
from sklearn.kernel_ridge import KernelRidge
from sklearn.semi_supervised.label_propagation import LabelPropagation
from sklearn.semi_supervised.label_propagation import LabelSpreading
from sklearn.linear_model.least_angle import Lars
from sklearn.linear_model.coordinate_descent import Lasso
from sklearn.linear_model.least_angle import LassoLars
from sklearn.linear_model.least_angle import LassoLarsIC
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.linear_model.base import LinearRegression
from sklearn.svm.classes import LinearSVC
from sklearn.svm.classes import LinearSVR
from sklearn.linear_model.logistic import LogisticRegression
from sklearn.neural_network.multilayer_perceptron import MLPClassifier
from sklearn.neural_network.multilayer_perceptron import MLPRegressor
from sklearn.linear_model.coordinate_descent import MultiTaskElasticNet
from sklearn.linear_model.coordinate_descent import MultiTaskLasso
from sklearn.naive_bayes import MultinomialNB
from sklearn.neighbors.nearest_centroid import NearestCentroid
from sklearn.svm.classes import NuSVC
from sklearn.svm.classes import NuSVR
from sklearn.linear_model.omp import OrthogonalMatchingPursuit
from sklearn.cross_decomposition.pls_ import PLSCanonical
from sklearn.cross_decomposition.pls_ import PLSRegression
from sklearn.linear_model.passive_aggressive import PassiveAggressiveClassifier
from sklearn.linear_model.passive_aggressive import PassiveAggressiveRegressor
from sklearn.linear_model.perceptron import Perceptron
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.linear_model.ransac import RANSACRegressor
from sklearn.neighbors.classification import RadiusNeighborsClassifier
from sklearn.neighbors.regression import RadiusNeighborsRegressor
from sklearn.linear_model.ridge import Ridge
from sklearn.linear_model.ridge import RidgeClassifier
from sklearn.linear_model.stochastic_gradient import SGDClassifier
from sklearn.linear_model.stochastic_gradient import SGDRegressor
from sklearn.svm.classes import SVC
from sklearn.svm.classes import SVR
from sklearn.linear_model.theil_sen import TheilSenRegressor

In [4]:
mlts = [MultiOutputRegressor, ClassifierChain]
bss = [Ridge, LinearSVR, GaussianProcessRegressor, PassiveAggressiveRegressor, KNeighborsRegressor, MLPRegressor, HuberRegressor, SGDRegressor, LassoLarsIC, ExtraTreeRegressor, ElasticNet, DecisionTreeRegressor, Perceptron, OrthogonalMatchingPursuit, BernoulliNB]

In [5]:
tfidf = pandas.read_csv('07 NM PT TFIDF.csv', delimiter=',', encoding='latin-1').fillna('')
tfidf[:3]

____________________  _____________________  _____________________ follow  \
0                   0.0                    0.0                           0.0   
1                   0.0                    0.0                           0.0   
2                   0.0                    0.0                           0.0   

   able  acceptable  accessory  accessory require  accidentally  accommodate  \
0   0.0         0.0        0.0                0.0           0.0          0.0   
1   0.0         0.0        0.0                0.0           0.0          0.0   
2   0.0         0.0        0.0                0.0           0.0          0.0   

   accompany   ...    work  would  write  writer  wrong  wrong patient  \
0        0.0   ...     0.0    0.0    0.0     0.0    0.0            0.0   
1        0.0   ...     0.0    0.0    0.0     0.0    0.0            0.0   
2        0.0   ...     0.0    0.0    0.0     0.0    0.0            0.0   

   wrong tattoo  xrt  xrt date  xrt pt  
0           0.0  0.0       0.0     0.0  
1           0.0  0.0       0.0     0.0  
2           0.0  0.0       0.0     0.0  

[3 rows x 1891 columns]

In [6]:
ohe = pandas.read_csv('07 NM PT OHE.csv', delimiter=',', encoding='latin-1').fillna('')
ohe[:3]

Wrong target or OAR contours  \
0                             0   
1                             0   
2                             0   

   Treatment plan acceptable but not physically deliverable  \
0                                                  0          
1                                                  0          
2                                                  0          

   Excess imaging dose   \
0                     0   
1                     0   
2                     0   

   Failure to perform on-treatment imaging as per instructions  Wrong patient  \
0                                                  0                        0   
1                                                  0                        0   
2                                                  0                        0   

   Fall or other patient injury or medical condition  \
0                                                  0   
1                                                  0   
2                                                  0   

   Wrong anatomical site (excluding laterality)  \
0                                             0   
1                                             0   
2                                             0   

   Treatment plan (isodose distribution) unacceptable  \
0                                                  0    
1                                                  0    
2                                                  0    

   Systematic hardware/software (including dose-volume) error  \
0                                                  0            
1                                                  0            
2                                                  0            

   Interventional procedure error (Retired value)        ...          \
0                                               0        ...           
1                                               0        ...           
2                                               0        ...           

   Inappropriate or poorly informed decision to treat or plan  \
0                                                  0            
1                                                  0            
2                                                  0            

   Inadequate coordination of combined modality care  \
0                                                  0   
1                                                  0   
2                                                  0   

   Wrong, missing, mislabeled, or damaged treatment accessories  \
0                                                  1              
1                                                  0              
2                                                  0              

   Wrong side (laterality)  \
0                        0   
1                        0   
2                        0   

   Wrong prescription dose-fractionation or calculation error  \
0                                                  0            
1                                                  0            
2                                                  0            

   Radiation therapy scheduling error  \
0                                   0   
1                                   1   
2                                   1   

   Treatment not delivered - personnel/hardware/software failure  \
0                                                  0               
1                                                  0               
2                                                  0               

   Wrong patient position, setup point, or shift  Wrong planning margins  \
0                                              0                       0   
1                                              0                       0   
2                                              0                       0   

   Allergic reaction  
0                  0  
1                  0  
2                

In [7]:
def lowest_correct(trues, preds):
    num_of_options = len(trues)
    drop_down_options = list(reversed(numpy.argsort(preds)))
    correct_options = [i for i in range(num_of_options) if trues[i]==1]
    return max([drop_down_options.index(correct_option) for correct_option in correct_options]) + 1
def average_lowest_correct(list_of_trues, list_of_preds):
    length = len(list_of_trues)
    return numpy.mean([lowest_correct(list(list_of_trues.iloc[i]), list(list_of_preds[i])) for i in range(length)])

## Default

In [8]:
def deadline(timeout, *args):
    def decorate(f):
        def handler(signum, frame):
            raise Exception

        def new_f(*args):
            signal.signal(signal.SIGALRM, handler)
            signal.alarm(timeout)
            return f(*args)
            signal.alarm(0)

        new_f.__name__ = f.__name__
        return new_f
    return decorate

In [9]:
@deadline(500)
def cv_ensemble(model, xs, ys):
    temp = cross_validate(model, xs, ys, return_train_score=True, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2)
    return numpy.mean(temp['test_score']).round(2)

In [10]:
cv_ensemble(MultiOutputRegressor(SGDRegressor()), tfidf, ohe)

4.35

## Grid Search

In [10]:
def prn_rsl(gsc):
    prms = gsc.cv_results_['params']
    tsts = (gsc.cv_results_['split0_test_score']+gsc.cv_results_['split1_test_score'])/2
    for i in numpy.argsort(tsts):
        print(prms[i], tsts[i])

In [11]:
ppl = Pipeline([('mb', MultiOutputRegressor(SGDRegressor()))])

In [13]:
grd = {
    'mb__estimator__loss': ['squared_loss', 'huber', 'epsilon_insensitive', 'squared_epsilon_insensitive'],
    'mb__estimator__penalty': ['none', 'l2', 'l1', 'elasticnet'],
    'mb__estimator__fit_intercept': [True, False]
}
gsc = GridSearchCV(estimator=ppl, param_grid=grd, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2, verbose=0, error_score=0)
gsc.fit(tfidf, ohe)
prn_rsl(gsc)

{'mb__estimator__fit_intercept': False, 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'elasticnet'} 3.1558944587235116
{'mb__estimator__fit_intercept': False, 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'l2'} 3.157487373510434
{'mb__estimator__fit_intercept': False, 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'none'} 3.1585455721911675
{'mb__estimator__fit_intercept': False, 'mb__estimator__loss': 'squared_epsilon_insensitive', 'mb__estimator__penalty': 'l2'} 3.164370158347862
{'mb__estimator__fit_intercept': False, 'mb__estimator__loss': 'huber', 'mb__estimator__penalty': 'none'} 3.169138792530151
{'mb__estimator__fit_intercept': False, 'mb__estimator__loss': 'huber', 'mb__estimator__penalty': 'l2'} 3.177088201197045
{'mb__estimator__fit_intercept': False, 'mb__estimator__loss': 'squared_epsilon_insensitive', 'mb__estimator__penalty': 'none'} 3.1834503118428383
{'mb__estimator__fit_intercept': False, 'mb__estimator__loss': '

In [14]:
grd = {
    'mb__estimator__loss': ['squared_loss'],
    'mb__estimator__penalty': ['elasticnet'],
    'mb__estimator__fit_intercept': [False],
    'mb__estimator__learning_rate': ['constant', 'optimal', 'invscaling'],
    'mb__estimator__average': [True, False]
}
gsc = GridSearchCV(estimator=ppl, param_grid=grd, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2, verbose=0, error_score=0)
gsc.fit(tfidf, ohe)
prn_rsl(gsc)

{'mb__estimator__average': False, 'mb__estimator__fit_intercept': False, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'elasticnet'} 2.9714426775347342
{'mb__estimator__average': True, 'mb__estimator__fit_intercept': False, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'elasticnet'} 3.0695012985962578
{'mb__estimator__average': False, 'mb__estimator__fit_intercept': False, 'mb__estimator__learning_rate': 'invscaling', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'elasticnet'} 3.1532410985495263
{'mb__estimator__average': True, 'mb__estimator__fit_intercept': False, 'mb__estimator__learning_rate': 'invscaling', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'elasticnet'} 3.173922591980157
{'mb__estimator__average': True, 'mb__estimator__fit_intercept': False, 'mb__estimator__learning_rate': 'optimal', 'mb__estimator__loss': '

In [16]:
grd = {
    'mb__estimator__loss': ['squared_loss'],
    'mb__estimator__penalty': ['elasticnet'],
    'mb__estimator__fit_intercept': [False],
    'mb__estimator__learning_rate': ['constant'],
    'mb__estimator__alpha': [0.1**i for i in range(2, 7)],
    'mb__estimator__l1_ratio': numpy.linspace(0.25, 1, 5)
}
gsc = GridSearchCV(estimator=ppl, param_grid=grd, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2, verbose=0, error_score=0)
gsc.fit(tfidf, ohe)
prn_rsl(gsc)

{'mb__estimator__alpha': 1.0000000000000004e-06, 'mb__estimator__fit_intercept': False, 'mb__estimator__l1_ratio': 0.8125, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'elasticnet'} 2.9629607994679796
{'mb__estimator__alpha': 1.0000000000000003e-05, 'mb__estimator__fit_intercept': False, 'mb__estimator__l1_ratio': 0.8125, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'elasticnet'} 2.9661415739525854
{'mb__estimator__alpha': 1.0000000000000004e-06, 'mb__estimator__fit_intercept': False, 'mb__estimator__l1_ratio': 1.0, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__penalty': 'elasticnet'} 2.966671796646117
{'mb__estimator__alpha': 1.0000000000000003e-05, 'mb__estimator__fit_intercept': False, 'mb__estimator__l1_ratio': 0.4375, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb

In [17]:
grd = {
    'mb__estimator__loss': ['squared_loss'],
    'mb__estimator__penalty': ['elasticnet'],
    'mb__estimator__fit_intercept': [False],
    'mb__estimator__learning_rate': ['constant'],
    'mb__estimator__alpha': [0.000001],
    'mb__estimator__l1_ratio': [0.8125],
    'mb__estimator__max_iter': [5, 50, 500],
    
}
gsc = GridSearchCV(estimator=ppl, param_grid=grd, scoring=make_scorer(average_lowest_correct), n_jobs=1, cv=2, verbose=0, error_score=0)
gsc.fit(tfidf, ohe)
prn_rsl(gsc)

{'mb__estimator__alpha': 1e-06, 'mb__estimator__fit_intercept': False, 'mb__estimator__l1_ratio': 0.8125, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__max_iter': 5, 'mb__estimator__penalty': 'elasticnet'} 2.9687938107734064
{'mb__estimator__alpha': 1e-06, 'mb__estimator__fit_intercept': False, 'mb__estimator__l1_ratio': 0.8125, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__max_iter': 50, 'mb__estimator__penalty': 'elasticnet'} 3.0900451812642666
{'mb__estimator__alpha': 1e-06, 'mb__estimator__fit_intercept': False, 'mb__estimator__l1_ratio': 0.8125, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__max_iter': 500, 'mb__estimator__penalty': 'elasticnet'} 4.007939298488416


### Default
4.35
### Tuned
{'mb__estimator__alpha': 1e-06, 'mb__estimator__fit_intercept': False, 'mb__estimator__l1_ratio': 0.8125, 'mb__estimator__learning_rate': 'constant', 'mb__estimator__loss': 'squared_loss', 'mb__estimator__max_iter': 5, 'mb__estimator__penalty': 'elasticnet'} 

2.97